# Measure foci count and nucleus FISH intensities

- scripts to qiantify following parameters
    1. Foci counts per nuclei
    2. Foci intensities per nuclei (total integrated intensities & avg intensities per focus)
    3. Nucleus intensities


## 1. Prepare sum intensity projections

Raw images are (1) split for FISH channel; (2) rolling-ball background subtracted slice-by-slice along the z with user-defiend radius; (3) sum z-projected; and (4) saved with user-defined suffix at a user-defined directory. 

In [1]:
import sys
from concurrent.futures import ThreadPoolExecutor, as_completed
from pathlib import Path

sys.path.insert(0, "src")

import numpy as np
import tifffile

from imagej_rolling_ball import rolling_ball_background, subtract_background_image

# --- User-defined -------------------------------------------------------------------
INPUT_DIR = "/Volumes/Jeff-exFAT/Marko_HIV/Marko_HIV_rawsorted"
OUTPUT_DIR = "/Volumes/Jeff-exFAT/Marko_HIV/Marko_HIV_ch0bgs100sum"
PATTERN = "*.tif"
CHANNEL_AXIS = 1           # 0-based, in the raw (Z, C, Y, X) order
CHANNEL_INDEX = 0          # 0-based FISH channel
Z_AXIS = 0                 # 0-based Z axis after the channel axis is removed
ROLLING_BALL_RADIUS = 100  # pixels
SUFFIX = "_ch0bgs100sum"
WORKERS = 4                # images processed in parallel; ~1.5 GB RAM each
OVERWRITE = False          # False: skip images whose output already exists


# --- Split channel -> rolling ball per slice -> sum Z-project -> save --------------
def process(path, ref_shape):
    """Returns a list of reasons the file was skipped (empty if it was saved)."""
    raw = tifffile.imread(path)
    reasons = []
    if raw.ndim != len(ref_shape):
        reasons.append(f"ndim {raw.ndim} != {len(ref_shape)} (shape {raw.shape} vs {ref_shape})")
    elif raw.shape[CHANNEL_AXIS] != ref_shape[CHANNEL_AXIS]:
        reasons.append(f"{raw.shape[CHANNEL_AXIS]} channels != {ref_shape[CHANNEL_AXIS]} (shape {raw.shape} vs {ref_shape})")
    if raw.ndim > CHANNEL_AXIS and CHANNEL_INDEX >= raw.shape[CHANNEL_AXIS]:
        reasons.append(f"CHANNEL_INDEX {CHANNEL_INDEX} out of range for {raw.shape[CHANNEL_AXIS]} channels")
    if reasons:
        return reasons

    stack = np.moveaxis(np.take(raw, CHANNEL_INDEX, axis=CHANNEL_AXIS), Z_AXIS, 0)
    subtracted, _ = subtract_background_image(stack, rolling_ball_background(stack, ROLLING_BALL_RADIUS))
    sum_projection = subtracted.sum(axis=0).astype(np.float32)  # float32: a sum can overflow uint16
    tifffile.imwrite(out_dir / f"{path.stem}{SUFFIX}.tif", sum_projection)
    return []


out_dir = Path(OUTPUT_DIR)
out_dir.mkdir(parents=True, exist_ok=True)

paths = sorted(Path(INPUT_DIR).glob(PATTERN))
ref_shape = tifffile.TiffFile(paths[0]).series[0].shape  # the others must match its ndim and channel count
todo = [p for p in paths if OVERWRITE or not (out_dir / f"{p.stem}{SUFFIX}.tif").exists()]
print(f"{len(paths)} files matching {PATTERN}, {len(paths) - len(todo)} already done, "
      f"{len(todo)} to process on {WORKERS} workers (reference shape {ref_shape})")

skipped = []  # (file name, reasons)
with ThreadPoolExecutor(WORKERS) as pool:  # scipy.ndimage releases the GIL, so threads run in parallel
    futures = {pool.submit(process, p, ref_shape): p for p in todo}
    for i, future in enumerate(as_completed(futures), 1):
        path = futures[future]
        try:
            reasons = future.result()
        except Exception as err:  # unreadable file etc.: log it and move on
            reasons = [f"{type(err).__name__}: {err}"]
        if reasons:
            skipped.append((path.name, reasons))
        print(f"[{i}/{len(todo)}] {'SKIP ' if reasons else ''}{path.name}"
              + (f": {'; '.join(reasons)}" if reasons else ""))

# --- Log skipped files ---------------------------------------------------------------
log_path = out_dir / f"skipped{SUFFIX}.log"
log_path.write_text("".join(f"{name}\t{'; '.join(reasons)}\n" for name, reasons in sorted(skipped)))
print(f"done: {len(todo) - len(skipped)} saved, {len(skipped)} skipped -> {log_path}")

585 files matching *.tif, 0 already done, 585 to process on 4 workers (reference shape (11, 4, 2048, 2048))
[1/585] 080623_CPEB4_Dox_INF-24hpi_001.tif
[2/585] 080623_CPEB4_Dox_INF-24hpi_002.tif
[3/585] 080623_CPEB4_Dox_INF-24hpi_004.tif
[4/585] 080623_CPEB4_Dox_INF-24hpi_003.tif
[5/585] 080623_CPEB4_Dox_INF-24hpi_005.tif
[6/585] 080623_CPEB4_Dox_INF-24hpi_007.tif
[7/585] 080623_CPEB4_Dox_INF-24hpi_008.tif
[8/585] 080623_CPEB4_Dox_INF-24hpi_006.tif
[9/585] 080623_CPEB4_Dox_INF-24hpi_009.tif
[10/585] 080623_CPEB4_Dox_INF-24hpi_010.tif
[11/585] 080623_CPEB4_Dox_INF-48hpi_001.tif
[12/585] 080623_CPEB4_Dox_INF-48hpi_002.tif
[13/585] 080623_CPEB4_Dox_INF-48hpi_003.tif
[14/585] 080623_CPEB4_Dox_INF-48hpi_004.tif
[15/585] 080623_CPEB4_Dox_INF-48hpi_005.tif
[16/585] 080623_CPEB4_Dox_INF-48hpi_006.tif
[17/585] 080623_CPEB4_Dox_INF-48hpi_007.tif
[18/585] 080623_CPEB4_Dox_INF-48hpi_008.tif
[19/585] 080623_CPEB4_Dox_INF-48hpi_009.tif
[20/585] 080623_CPEB4_Dox_INF-48hpi_010.tif
[21/585] 080623_CPEB4

## 2. Measure foci count and intensities per nucleus, normalised to Mock

Two cells: **2a** measures every image (slow); **2b** normalises to Mock nuclei, filters foci and saves the tables (fast), so the baseline and filter settings can be changed and 2b re-run without re-measuring.

Inputs (2D, same shape), matched by basename (the raw file stem) from three user-defined directories:
- **nucleus label image**: `{basename}{LABEL_SUFFIX}.tif` (cellpose output). The label value is the nucleus id (`nucleus-id`).
- **foci mask**: `{basename}{FOCI_SUFFIX}.tif`. Any non-zero pixel is foreground.
- **sum intensity image**: `{basename}{SUM_SUFFIX}.tif` (section 1 output).

Rules:
- **Border nuclei**: nuclei touching the image sides are discarded. The remaining ids are kept as they are in the label image, never renumbered, so they can be cross-referenced.
- **Foci**: 4-connected blobs of the foci mask (diagonal-only contact = separate foci), given an arbitrary per-image `foci-index` that is not meant for cross-referencing.
- **Assignment**: each whole focus goes to the nucleus it overlaps most. Any overlap counts, even one pixel. Ties go to the lowest nucleus id and are logged. The majority is decided among *all* nuclei, so a focus won by a discarded border nucleus is dropped, not handed to a neighbour.
- **Foci inclusion** (`FOCI_INCLUSION`) decides which assigned foci are counted:
    - `"encapsulated"` (default): a focus is kept only if every one of its pixels lies inside its assigned nucleus. Foci that reach outside the nucleus, or into a neighbouring nucleus, are discarded whole.
    - `"overlapping"`: every assigned focus is kept, including those reaching outside the nucleus (flagged by `foci-extends-outside-nucleus`).
- **Foci measurements** use the whole blob (in `"encapsulated"` mode this is the same as the pixels inside the nucleus).
- Foci that overlap no nucleus are ignored.
- **Minimum foci area**: foci with area < `MIN_FOCI_AREA` pixels are removed before assignment, so they are neither counted nor able to win a nucleus. No other size or intensity filters are applied: foci masks and nuclei are otherwise taken as segmented.
- Nuclei with zero foci are **kept**: in the per-focus table as one row with `has-foci` = False and empty foci fields, and in the per-nucleus table with `foci-count` = 0, totals = 0, means empty.
- Areas are in pixels. Intensities are sums of sum-projection pixel values. **"raw"** means measured on the section 1 image, i.e. already rolling-ball background subtracted, before Mock normalisation.

Conditions (parsed from the basename `{replicate}_{rbp}_{induction}_{hiv-infection}_{field}`, e.g. `080623_CPEB4_Dox_INF-24hpi_001`):
- `replicate`: the 6-digit date (biological replicate); `rbp`: CPEB4 / GFP / MARF1; `induction`: Dox / NoDox; `hiv-infection`: INF-24hpi / INF-48hpi / Mock; `field`: the image number.
- A name that does not match, or has a value outside these lists, is logged as `ERROR` and the image is skipped.
- All are text. `replicate` and `field` keep their leading zeros, so read them back as strings (e.g. pandas `dtype={"replicate": str, "field": str}`); otherwise `080623` becomes `80623`.

Mock normalisation (2b), to remove nuclear autofluorescence and non-specific probe signal:
1. Nuclei are grouped by `BASELINE_GROUP_BY` (e.g. per `replicate`).
2. In each group, the **Mock** nuclei (all kept Mock nuclei, with or without foci) give a background per pixel from their densities (raw integrated intensity / area), summarised by `BASELINE_STATISTIC`:
    - `"median"`: median of the per-nucleus densities (robust to outlier nuclei)
    - `"mean"`: mean of the per-nucleus densities (each nucleus weighted equally)
    - `"pooled-mean"`: sum of intensities / sum of areas (weighted by area)
3. Every nucleus and focus in the group (Mock and infected) gets `normalised = raw - area x background per pixel`.
- Negative values are expected (Mock nuclei average about 0) and are kept, not clipped.
- Every group must have Mock nuclei, otherwise 2b stops with an error naming the groups.
- A focus's normalised intensity is its specific signal, including any diffuse nucleoplasmic HIV RNA under it; it is not background-corrected against the surrounding nucleoplasm.
- Only an additive offset is removed. Multiplicative day-to-day differences (laser power, probe batch) remain, so treat `replicate` as a blocking factor in statistics.

Mock foci filter (2b, after normalisation). Foci masks come from a per-image auto-threshold, which splits noise into foci in images without real foci, so Mock nuclei (true negatives) show false foci. `MOCK_FOCI_FILTER` decides which detected foci are **counted** (`foci-counted`); uncounted foci keep their rows but are left out of the per-nucleus counts and totals:
- `"cutoff"` (default, Mock-driven): per `FOCI_CUTOFF_GROUP_BY` group (default per `replicate`), cutoff = the `MOCK_FOCI_CUTOFF_PERCENTILE` (default 99) percentile of the **normalised integrated intensity** of the foci detected in Mock nuclei. In **every** condition, a focus is counted only if its normalised integrated intensity > cutoff. About (100 - percentile)% of Mock foci stay counted: that is the false-positive rate, reported rather than forced to 0. A group whose Mock nuclei have no foci gets no cutoff, and all its foci are counted.
- `"keep"`: every detected focus is counted.
- `"drop"`: foci in Mock nuclei are not counted; all others are. Infected images keep their noise foci, so foci counts are not comparable between Mock and infected.
- Every group must have Mock nuclei, otherwise 2b stops with an error.
- If noise and real foci have similar intensities, no cutoff separates them: check the distribution plot (output 5).

Reference scaling (2b, optional, `REFERENCE_CONDITION`; `None` skips it). Puts replicates on one scale for pooled plots:
- Per replicate, every normalised intensity column is divided by its mean in the reference condition (default GFP NoDox INF-48hpi, present in every replicate), giving `{column}-relative-to-reference`. The reference averages 1 in every replicate.
- Scaled: `nucleus-fish-normalised-integrated-intensity`, `foci-fish-normalised-integrated-intensity` (per-focus table; reference mean over counted reference foci), `total-foci-fish-normalised-integrated-intensity` and `mean-fish-normalised-integrated-intensity-per-foci` (per-nucleus table; reference mean over reference nuclei, the latter over those with counted foci). Foci counts are not scaled.
- Use an infected reference: Mock is about 0 after normalisation. A replicate without reference nuclei or counted foci stops 2b with an error.
- The factor absorbs imaging/labelling scale and day-to-day infection efficiency. The reference loses its between-replicate spread, so run statistics on replicate-level values (paired / ratio tests on replicate means, or a mixed model with replicate as a random effect), not on pooled nuclei.

Outputs (in `OUTPUT_DIR`). `{statistic}` is `BASELINE_STATISTIC`; `{p}` is `MOCK_FOCI_CUTOFF_PERCENTILE`; columns marked *(cutoff)* are present only with `MOCK_FOCI_FILTER = "cutoff"`; with a `REFERENCE_CONDITION`, each scaled column is followed by its `...-relative-to-reference` column:
1. `foci_intensity_quantification.csv`: the full table. One row per detected focus, plus one row per kept nucleus without foci (`has-foci` = False, foci fields empty). Nucleus columns repeat across the foci of a nucleus. Count foci with `foci-counted` (or `has-foci` for all detected), not rows.
    - `image-filename-basename`, `replicate`, `rbp`, `induction`, `hiv-infection`, `field`, `nucleus-id`, `nucleus-area`, `mock-nuclear-fish-background-per-pixel-{statistic}`, `nucleus-fish-raw-integrated-intensity`, `nucleus-fish-normalised-integrated-intensity`, `has-foci`, `foci-index`, `foci-area`, `foci-fish-raw-integrated-intensity`, `foci-fish-normalised-integrated-intensity`, `mock-foci-cutoff-p{p}-normalised-integrated-intensity` *(cutoff)*, `foci-counted` (empty for nuclei without foci), `foci-max-intensity` (raw), `foci-centroid-x`, `foci-centroid-y` (pixels; x = column, y = row), `foci-extends-outside-nucleus`
2. `foci_intensity_quantification_summarised-per-nucleus.csv`: one row per kept nucleus, summarised from table 1. Foci counts, totals and means use counted foci only.
    - `image-filename-basename`, `replicate`, `rbp`, `induction`, `hiv-infection`, `field`, `nucleus-id`, `nucleus-area`, `mock-nuclear-fish-background-per-pixel-{statistic}`, `mock-foci-cutoff-p{p}-normalised-integrated-intensity` *(cutoff)*, `nucleus-fish-raw-integrated-intensity`, `nucleus-fish-normalised-integrated-intensity`, `foci-detected-count`, `foci-count` (counted), `total-foci-area`, `total-foci-fish-raw-integrated-intensity`, `total-foci-fish-normalised-integrated-intensity`, `mean-fish-raw-integrated-intensity-per-foci`, `mean-fish-normalised-integrated-intensity-per-foci` (= total / count)
3. `foci_intensity_quantification_mock-baseline.csv`: one row per baseline group.
    - the `BASELINE_GROUP_BY` columns, `n-mock-nuclei`, `n-mock-images`, `median`, `mean`, `pooled-mean` (background per pixel by each statistic), `sd` (sample SD of the per-nucleus densities), `statistic-used`
4. `foci_intensity_quantification_mock-foci-filter.csv`: one row per foci filter group.
    - the `FOCI_CUTOFF_GROUP_BY` columns, `mock-foci-filter`, `percentile`, `cutoff` (both empty unless `"cutoff"`), `n-mock-nuclei`, `n-mock-foci-detected`, `n-mock-foci-counted`, `mock-counted-foci-per-nucleus` (false-positive rate), `n-foci-detected`, `n-foci-counted` (all conditions)
5. `foci_intensity_quantification_foci-intensity-distribution.png`: per filter group, the distribution of focus normalised integrated intensity (log10) for every detected focus, one line per `hiv-infection`, each scaled to its own number of foci; the cutoff as a dashed line. Foci with normalised intensity <= 0 cannot be shown on the log axis and are counted in the legend. Use it to check that Mock (noise) and infected foci separate.
6. `foci_intensity_quantification_reference-factors.csv` (only with a `REFERENCE_CONDITION`): one row per replicate.
    - `replicate`, `rbp`, `induction`, `hiv-infection` (the reference condition), `n-reference-nuclei`, `n-reference-foci-counted`, `{column}-reference-mean` for each scaled column
7. `foci_intensity_quantification.log` (written by 2a): tab-separated `image  level  message`.
    - `ERROR`: missing or mismatched inputs, unreadable files, basenames that cannot be parsed.
    - `WARNING`: sum images with no label image; images with no nuclei or only border nuclei.
    - `OUTLIER`: foci overlapping more than one nucleus (with pixel counts and the winner).
    - `INFO`: per-image nucleus and foci counts (nuclei kept / border; foci counted / below the minimum area / on border nuclei / outside nuclei / not encapsulated (`"encapsulated"` only)).
8. `foci_intensity_quantification_params.json`: all settings, the run times, file and row counts and package versions.
9. `qc/{basename}_nucleus-foci-qc.png` (written by 2a): sum projection at 1:1 pixels. Kept nuclei in cyan with their id, border nuclei in grey, counted foci in magenta, ignored foci in yellow (too small, on a border nucleus, outside nuclei, or not encapsulated). Drawn before the Mock foci filter, so foci later not counted are still magenta.

### Planned, not yet implemented
To be added on top of the Mock-normalised intensities:
- mean intensity (integrated / area) for nuclei and foci
- nucleoplasm: the nucleus minus its foci (area, integrated and mean intensity)
- fraction of nuclear signal in (counted) foci = `total-foci-fish-normalised-integrated-intensity` / `nucleus-fish-normalised-integrated-intensity`. Unstable when the normalised nucleus intensity is near 0 or negative (low-signal nuclei). With `FOCI_INCLUSION = "overlapping"`, foci are measured as whole blobs, so pixels outside the nucleus are included in the numerator; `"encapsulated"` avoids this.

In [4]:
# --- 2a. Measure every image ------------------------------------------------------------
import json
import math
import sys
import traceback
from concurrent.futures import ThreadPoolExecutor, as_completed
from datetime import datetime
from importlib.metadata import version
from pathlib import Path

sys.path.insert(0, "src")

import numpy as np
import tifffile

from nucleus_foci_quant import measure_image, parse_basename, qc_figure

# --- User-defined -------------------------------------------------------------------
LABEL_DIR = "/Volumes/Jeff-exFAT/Marko_HIV/Marko_HIV_cpnuclmask"  # cellpose output
FOCI_DIR = "/Volumes/Jeff-exFAT/Marko_HIV/Marko_HIV_focimask"
SUM_DIR = "/Volumes/Jeff-exFAT/Marko_HIV/Marko_HIV_ch0bgs100sum"
OUTPUT_DIR = "/Volumes/Jeff-exFAT/Marko_HIV/Marko_HIV_foci-quantification"
LABEL_SUFFIX = "__cpnuclmask"   # {basename}{LABEL_SUFFIX}.tif
FOCI_SUFFIX = "_focimask"    # {basename}{FOCI_SUFFIX}.tif
SUM_SUFFIX = "_ch0bgs100sum" # {basename}{SUM_SUFFIX}.tif
MIN_FOCI_AREA = 9            # px; foci with area < this are removed (0 keeps all)
FOCI_INCLUSION = "encapsulated"  # "encapsulated": keep only foci wholly inside the nucleus
                                 # "overlapping": also keep foci reaching outside the nucleus
SAVE_QC = True               # one PNG per image in OUTPUT_DIR/qc
WORKERS = 6


# --- Measure one image ----------------------------------------------------------------
def process(basename):
    """Returns (rows, messages). Raises on unparseable basenames and unreadable/missing inputs."""
    conditions = parse_basename(basename)
    paths = {
        "label": Path(LABEL_DIR) / f"{basename}{LABEL_SUFFIX}.tif",
        "foci": Path(FOCI_DIR) / f"{basename}{FOCI_SUFFIX}.tif",
        "sum": Path(SUM_DIR) / f"{basename}{SUM_SUFFIX}.tif",
    }
    missing = [f"{kind} ({p.name})" for kind, p in paths.items() if not p.exists()]
    if missing:
        raise FileNotFoundError(f"missing {', '.join(missing)}")
    labels, foci, intensity = (tifffile.imread(paths[k]) for k in ("label", "foci", "sum"))
    rows, messages, foci_lab, assigned = measure_image(
        basename, labels, foci, intensity, min_foci_area=MIN_FOCI_AREA,
        foci_inclusion=FOCI_INCLUSION)
    if SAVE_QC:
        qc_figure(basename, labels, intensity, foci_lab, assigned).savefig(
            qc_dir / f"{basename}_nucleus-foci-qc.png")
    return [{**r, **conditions} for r in rows], messages


out_dir = Path(OUTPUT_DIR)
qc_dir = out_dir / "qc"
qc_dir.mkdir(parents=True, exist_ok=True)
started = datetime.now()

label_paths = sorted(Path(LABEL_DIR).glob(f"*{LABEL_SUFFIX}.tif"))
basenames = [p.name.removesuffix(f"{LABEL_SUFFIX}.tif") for p in label_paths]
log = []  # (basename, level, message)
for p in sorted(Path(SUM_DIR).glob(f"*{SUM_SUFFIX}.tif")):
    b = p.name.removesuffix(f"{SUM_SUFFIX}.tif")
    if b not in set(basenames):
        log.append((b, "WARNING", "sum intensity image has no nucleus label image"))
print(f"{len(basenames)} label images matching *{LABEL_SUFFIX}.tif, processing on {WORKERS} workers")

all_rows, n_failed = [], 0  # raw per-focus rows (incl. one per nucleus without foci)
with ThreadPoolExecutor(WORKERS) as pool:
    futures = {pool.submit(process, b): b for b in basenames}
    for i, future in enumerate(as_completed(futures), 1):
        b = futures[future]
        try:
            rows, messages = future.result()
        except Exception as err:  # log it and move on
            n_failed += 1
            log.append((b, "ERROR", f"{type(err).__name__}: {err}"))
            print(f"[{i}/{len(basenames)}] ERROR {b}: {err}")
            continue
        all_rows += rows
        log += [(b, level, msg) for level, msg in messages]
        n_outliers = sum(level == "OUTLIER" for level, _ in messages)
        print(f"[{i}/{len(basenames)}] {b}: {len({r['nucleus-id'] for r in rows})} nuclei, "
              f"{sum(r['has-foci'] for r in rows)} foci"
              + (f", {n_outliers} outlier(s)" if n_outliers else ""))

measured = datetime.now()
n_failed_basename = sum(level == "ERROR" and "basename" in msg for _, level, msg in log)

log_path = out_dir / "foci_intensity_quantification.log"
with open(log_path, "w") as fh:
    fh.write("image\tlevel\tmessage\n")
    fh.writelines(f"{b}\t{level}\t{msg}\n" for b, level, msg in sorted(log))

print(f"measured: {len(basenames) - n_failed} images, "
      f"{len({(r['image-filename-basename'], r['nucleus-id']) for r in all_rows})} nuclei, "
      f"{sum(r['has-foci'] for r in all_rows)} foci, {n_failed} failed "
      f"({n_failed_basename} unparseable basenames) -> log in {log_path}")

583 label images matching *__cpnuclmask.tif, processing on 3 workers
[1/583] 080623_CPEB4_Dox_INF-24hpi_002: 10 nuclei, 1 foci
[2/583] 080623_CPEB4_Dox_INF-24hpi_003: 15 nuclei, 2 foci
[3/583] 080623_CPEB4_Dox_INF-24hpi_001: 14 nuclei, 2 foci
[4/583] 080623_CPEB4_Dox_INF-24hpi_005: 7 nuclei, 1 foci
[5/583] 080623_CPEB4_Dox_INF-24hpi_004: 20 nuclei, 0 foci
[6/583] 080623_CPEB4_Dox_INF-24hpi_006: 13 nuclei, 3 foci
[7/583] 080623_CPEB4_Dox_INF-24hpi_007: 11 nuclei, 3 foci
[8/583] 080623_CPEB4_Dox_INF-24hpi_008: 21 nuclei, 4 foci
[9/583] 080623_CPEB4_Dox_INF-24hpi_009: 17 nuclei, 2 foci
[10/583] 080623_CPEB4_Dox_INF-24hpi_010: 17 nuclei, 5 foci, 1 outlier(s)
[11/583] 080623_CPEB4_Dox_INF-48hpi_001: 8 nuclei, 34 foci
[12/583] 080623_CPEB4_Dox_INF-48hpi_002: 14 nuclei, 42 foci
[13/583] 080623_CPEB4_Dox_INF-48hpi_003: 8 nuclei, 38 foci
[14/583] 080623_CPEB4_Dox_INF-48hpi_004: 17 nuclei, 12 foci
[15/583] 080623_CPEB4_Dox_INF-48hpi_005: 17 nuclei, 13 foci
[16/583] 080623_CPEB4_Dox_INF-48hpi_006

In [5]:
# --- 2b. Normalise to Mock nuclei, filter foci and save ------------------------------------
import csv
import importlib

import nucleus_foci_quant

importlib.reload(nucleus_foci_quant)  # pick up module edits made since 2a imported it
from nucleus_foci_quant import (apply_mock_foci_filter, baseline_column, baseline_columns,
                                foci_columns, foci_filter_columns, foci_intensity_figure,
                                mock_baselines, normalise, nucleus_columns, reference_columns,
                                reference_factors, scale_to_reference, summarise_per_nucleus)

# --- User-defined -------------------------------------------------------------------
BASELINE_GROUP_BY = ["replicate"]
# Parsed columns that define which Mock nuclei make up each baseline. Every nucleus (and
# its foci) is corrected with the baseline of the Mock nuclei sharing these values.
#   ["replicate"]                     one baseline per replicate, pooling the Mock nuclei of
#                                     all rbp/induction conditions
#   ["replicate", "rbp", "induction"] each condition corrected by its own matched Mock
# Allowed columns: "replicate", "rbp", "induction". Every group must contain Mock nuclei,
# otherwise this cell stops with an error.

BASELINE_STATISTIC = "median"
# How the Mock nuclei of a group give one background-per-pixel value:
#   "median"      median of per-nucleus (raw integrated intensity / area); robust to outlier nuclei
#   "mean"        mean of per-nucleus (raw integrated intensity / area); each nucleus weighted equally
#   "pooled-mean" sum(raw integrated intensity) / sum(area) over the group; weighted by area
# The baseline column is named mock-nuclear-fish-background-per-pixel-{BASELINE_STATISTIC}.

MOCK_FOCI_FILTER = "cutoff"
# Which detected foci are counted (flag `foci-counted`; uncounted foci keep their rows but
# are left out of the per-nucleus counts and totals). Per-image auto-thresholds split noise
# into foci in images with no real foci, so Mock (true negative) nuclei show false foci.
#   "cutoff" Mock-driven: per FOCI_CUTOFF_GROUP_BY group, cutoff = the
#            MOCK_FOCI_CUTOFF_PERCENTILE percentile of the normalised integrated intensity of
#            the foci detected in Mock nuclei. In every condition (Mock and infected) a focus
#            is counted only if its normalised integrated intensity > cutoff, so about
#            (100 - percentile)% of Mock foci remain: the false-positive rate.
#   "keep"   count every detected focus (no filter)
#   "drop"   count no foci in Mock nuclei; count every focus elsewhere. Infected images keep
#            their noise foci, so do not compare foci counts between Mock and infected.
MOCK_FOCI_CUTOFF_PERCENTILE = 99.5  # "cutoff" only; 0-100
FOCI_CUTOFF_GROUP_BY = ["replicate"]
# Parsed columns defining the cutoff groups, as for BASELINE_GROUP_BY ("replicate", "rbp",
# "induction"). Every group must contain Mock nuclei.

REFERENCE_CONDITION = {"rbp": "GFP", "induction": "NoDox", "hiv-infection": "INF-48hpi"}
# Optional (None to skip): put replicates on one scale for pooled plots. Per replicate,
# every normalised intensity column is divided by its mean in this condition, giving
# `{column}-relative-to-reference` columns (the reference averages 1 in every replicate).
# Pick an infected condition present in every replicate: Mock is ~0 after normalisation.
# Foci counts are not scaled. The factor absorbs imaging/labelling scale and day-to-day
# infection efficiency; the reference itself loses its between-replicate spread, so run
# statistics on replicate-level values.


# --- Mock baseline -> normalise (raw - area x baseline) -> filter foci -> summarise -------
def write_csv(path, columns, rows):
    with open(path, "w", newline="") as fh:
        writer = csv.DictWriter(fh, fieldnames=columns)
        writer.writeheader()
        writer.writerows({k: ("" if isinstance(v, float) and math.isnan(v) else v) for k, v in r.items()}
                         for r in rows)


cutoff_percentile = MOCK_FOCI_CUTOFF_PERCENTILE if MOCK_FOCI_FILTER == "cutoff" else None
baselines = mock_baselines(all_rows, BASELINE_GROUP_BY, BASELINE_STATISTIC)
normalised = normalise(all_rows, baselines, BASELINE_GROUP_BY, BASELINE_STATISTIC)
foci_table, foci_filter = apply_mock_foci_filter(normalised, FOCI_CUTOFF_GROUP_BY, MOCK_FOCI_FILTER,
                                                 MOCK_FOCI_CUTOFF_PERCENTILE)
foci_table.sort(key=lambda r: (r["image-filename-basename"], r["nucleus-id"], r["foci-index"] or 0))
nucleus_table = summarise_per_nucleus(foci_table, BASELINE_STATISTIC, cutoff_percentile)
relative = REFERENCE_CONDITION is not None
if relative:
    reference = reference_factors(foci_table, nucleus_table, REFERENCE_CONDITION)
    foci_table = scale_to_reference(foci_table, reference, "foci")
    nucleus_table = scale_to_reference(nucleus_table, reference, "nucleus")
for row in baselines.values():
    group = ", ".join(f"{c}={row[c]}" for c in BASELINE_GROUP_BY) or "all"
    print(f"baseline {group}: {row[BASELINE_STATISTIC]:.4g} per px ({BASELINE_STATISTIC}) from "
          f"{row['n-mock-nuclei']} Mock nuclei in {row['n-mock-images']} images")
for row in foci_filter.values():
    group = ", ".join(f"{c}={row[c]}" for c in FOCI_CUTOFF_GROUP_BY) or "all"
    print(f"foci filter {group} ({MOCK_FOCI_FILTER}"
          + (f", cutoff {row['cutoff']:.4g}" if MOCK_FOCI_FILTER == "cutoff" else "") + "): "
          f"{row['n-foci-counted']}/{row['n-foci-detected']} foci counted; Mock "
          f"{row['n-mock-foci-counted']}/{row['n-mock-foci-detected']} "
          f"({row['mock-counted-foci-per-nucleus']:.3g} per Mock nucleus)")
for row in (reference.values() if relative else []):
    print(f"reference replicate={row['replicate']}: {row['n-reference-nuclei']} nuclei, "
          f"{row['n-reference-foci-counted']} counted foci; mean total foci intensity "
          f"{row['total-foci-fish-normalised-integrated-intensity-reference-mean']:.4g}")

# --- Save tables and parameters -------------------------------------------------------
write_csv(out_dir / "foci_intensity_quantification.csv",
          foci_columns(BASELINE_STATISTIC, cutoff_percentile, relative), foci_table)
write_csv(out_dir / "foci_intensity_quantification_summarised-per-nucleus.csv",
          nucleus_columns(BASELINE_STATISTIC, cutoff_percentile, relative), nucleus_table)
reference_path = out_dir / "foci_intensity_quantification_reference-factors.csv"
if relative:
    write_csv(reference_path, reference_columns(), list(reference.values()))
else:
    reference_path.unlink(missing_ok=True)  # no stale factors from an earlier run
write_csv(out_dir / "foci_intensity_quantification_mock-baseline.csv",
          baseline_columns(BASELINE_GROUP_BY), list(baselines.values()))
write_csv(out_dir / "foci_intensity_quantification_mock-foci-filter.csv",
          foci_filter_columns(FOCI_CUTOFF_GROUP_BY), list(foci_filter.values()))
foci_intensity_figure(foci_table, foci_filter, FOCI_CUTOFF_GROUP_BY,
                      title=f"Detected foci; Mock foci filter: {MOCK_FOCI_FILTER}").savefig(
    out_dir / "foci_intensity_quantification_foci-intensity-distribution.png", dpi=150)

params = {
    "started": started.isoformat(timespec="seconds"),
    "measured": measured.isoformat(timespec="seconds"),
    "saved": datetime.now().isoformat(timespec="seconds"),
    "LABEL_DIR": LABEL_DIR, "FOCI_DIR": FOCI_DIR, "SUM_DIR": SUM_DIR, "OUTPUT_DIR": OUTPUT_DIR,
    "LABEL_SUFFIX": LABEL_SUFFIX, "FOCI_SUFFIX": FOCI_SUFFIX, "SUM_SUFFIX": SUM_SUFFIX,
    "MIN_FOCI_AREA": MIN_FOCI_AREA, "FOCI_INCLUSION": FOCI_INCLUSION, "SAVE_QC": SAVE_QC, "WORKERS": WORKERS,
    "BASELINE_GROUP_BY": BASELINE_GROUP_BY, "BASELINE_STATISTIC": BASELINE_STATISTIC,
    "MOCK_FOCI_FILTER": MOCK_FOCI_FILTER, "MOCK_FOCI_CUTOFF_PERCENTILE": MOCK_FOCI_CUTOFF_PERCENTILE,
    "FOCI_CUTOFF_GROUP_BY": FOCI_CUTOFF_GROUP_BY, "REFERENCE_CONDITION": REFERENCE_CONDITION,
    "rules": {
        "foci_connectivity": 4,
        "border_nuclei": "excluded; ids kept from the label image",
        "foci_assignment": "whole blob to the max-overlap nucleus among all nuclei; any overlap counts; "
                           "ties to lowest id; dropped if the winner is a border nucleus",
        "foci_inclusion": ("only foci with every pixel inside the assigned nucleus"
                           if FOCI_INCLUSION == "encapsulated"
                           else "all assigned foci, including those reaching outside the nucleus"),
        "foci_measurement": "whole blob, including pixels outside the nucleus",
        "size_filters": f"foci with area < {MIN_FOCI_AREA} px removed before assignment; no nucleus filters",
        "raw": "measured on the rolling-ball subtracted sum projection (section 1)",
        "normalisation": f"normalised = raw - area x {baseline_column(BASELINE_STATISTIC)}, the "
                         f"{BASELINE_STATISTIC} background per pixel of the Mock nuclei grouped by "
                         f"{BASELINE_GROUP_BY}; negative values kept",
        "mock_foci_filter": {
            "cutoff": f"foci counted if normalised integrated intensity > the p{MOCK_FOCI_CUTOFF_PERCENTILE:g} "
                      f"of Mock foci normalised integrated intensity, grouped by {FOCI_CUTOFF_GROUP_BY}; "
                      f"applied to all conditions",
            "keep": "all detected foci counted",
            "drop": "foci in Mock nuclei not counted; all others counted",
        }[MOCK_FOCI_FILTER],
        "reference": (f"per replicate, each normalised intensity column / its mean in {REFERENCE_CONDITION} "
                      f"-> *-relative-to-reference" if relative else "not applied"),
    },
    "images": {"label_images": len(basenames), "processed": len(basenames) - n_failed, "failed": n_failed,
               "unparseable_basenames": n_failed_basename},
    "rows": {"per-focus table": len(foci_table), "foci detected": sum(r["has-foci"] for r in foci_table),
             "foci counted": sum(bool(r["foci-counted"]) for r in foci_table),
             "nuclei": len(nucleus_table), "baseline groups": len(baselines),
             "foci filter groups": len(foci_filter)},
    "versions": {"python": sys.version.split()[0],
                 **{pkg: version(pkg) for pkg in ("numpy", "scipy", "scikit-image", "tifffile", "matplotlib")}},
}
(out_dir / "foci_intensity_quantification_params.json").write_text(json.dumps(params, indent=2))

print(f"saved: {len(nucleus_table)} nuclei, {params['rows']['foci counted']}/"
      f"{params['rows']['foci detected']} foci counted, {len(baselines)} baseline group(s) -> {out_dir}")

baseline replicate=080623: 9139 per px (median) from 1195 Mock nuclei in 66 images
baseline replicate=110723: 5711 per px (median) from 1613 Mock nuclei in 62 images
baseline replicate=160323: 4241 per px (median) from 841 Mock nuclei in 26 images
baseline replicate=310323: 4655 per px (median) from 1574 Mock nuclei in 54 images
foci filter replicate=080623 (cutoff, cutoff 4.905e+05): 3441/16852 foci counted; Mock 68/13464 (0.0569 per Mock nucleus)
foci filter replicate=110723 (cutoff, cutoff 5.216e+05): 6461/13010 foci counted; Mock 33/6542 (0.0205 per Mock nucleus)
foci filter replicate=160323 (cutoff, cutoff 5.541e+05): 1942/12053 foci counted; Mock 51/10152 (0.0606 per Mock nucleus)
foci filter replicate=310323 (cutoff, cutoff 2.824e+05): 4993/44663 foci counted; Mock 200/39868 (0.127 per Mock nucleus)
reference replicate=080623: 244 nuclei, 583 counted foci; mean total foci intensity 2.547e+07
reference replicate=110723: 238 nuclei, 451 counted foci; mean total foci intensity 1.36